File: src/main/java/ru/sicampus/bootcamp2026/util/SecurityUtils.java
package ru.sicampus.bootcamp2026.util;

import org.springframework.security.core.context.SecurityContextHolder;
import ru.sicampus.bootcamp2026.entity.User;

public final class SecurityUtils {

    private SecurityUtils() {}

    public static User getCurrentUser() {
        var auth = SecurityContextHolder.getContext().getAuthentication();

        if (auth == null || !auth.isAuthenticated()) {
            throw new IllegalStateException("user not found but it should be");
        }

        return (User) auth.getPrincipal();
    }
}

File: src/main/kotlin/ru/sicampus/bootcamp2026/security/SecurityUtils.kt
package ru.sicampus.bootcamp2026.security

import org.springframework.security.access.AccessDeniedException
import org.springframework.security.core.Authentication
import org.springframework.security.core.context.SecurityContextHolder
import ru.sicampus.bootcamp2026.entity.User
import ru.sicampus.bootcamp2026.entity.UserRole
import ru.sicampus.bootcamp2026.repository.UserRepository

object SecurityUtils {
    fun getCurrentUserEmail(): String {
        val authentication: Authentication? = SecurityContextHolder.getContext().authentication
        return authentication?.name
            ?: throw IllegalStateException("No authenticated user found")
    }

    fun getCurrentUser(userRepository: UserRepository): User {
        val email = getCurrentUserEmail()
        return userRepository.findByEmail(email)
            .orElseThrow { IllegalStateException("Current user not found") }
    }

    fun isAdmin(user: User): Boolean = user.role == UserRole.ADMIN

    fun requireOwnershipOrAdmin(user: User, resourceOwnerId: Long) {
        if (!isAdmin(user) && user.id != resourceOwnerId) {
            throw AccessDeniedException("Access denied")
        }
    }

    fun requireAdmin(user: User) {
        if (!isAdmin(user)) {
            throw AccessDeniedException("Access denied")
        }
    }
}
